api/krusty/component_test.go

```diff
@@ -0,0 +1,670 @@
+  annotations: {}
+  labels: {}
+  name: comp-my-configmap-dgf97tmg6h
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: comp-db
+spec:
+  type: Logical
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: comp-stub
+spec:
+  replicas: 1
+`,
+		},
+		// If a component sets a name prefix on a base, then that base can also be separately included
+		// without being affected by the component in another branch of the resource tree
+		"basic-component-with-repeated-base": {
+			input: []FileGen{writeTestBase, writeTestComponent, writeOverlayProd,
+				writeK("/app/repeated", `
+resources:
+- ../base
+- ../prod
+`),
+			},
+			runPath: "/app/repeated",
+			expectedOutput: `
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: storefront
+spec:
+  replicas: 1
+---
+apiVersion: v1
+data:
+  otherValue: "10"
+  testValue: "1"
+kind: ConfigMap
+metadata:
+  name: my-configmap-7k9t4h74f8
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: comp-storefront
+spec:
+  replicas: 3
+---
+apiVersion: v1
+data:
+  compValue: "5"
+  otherValue: "10"
+  testValue: "2"
+kind: ConfigMap
+metadata:
+  annotations: {}
+  labels: {}
+  name: comp-my-configmap-ct5bgtbccd
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: comp-db
+spec:
+  type: Logical
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: comp-stub
+spec:
+  replicas: 1
+`,
+		},
+		"applying-component-directly-should-be-same-as-kustomization": {
+			input: []FileGen{writeTestBase, writeTestComponent,
+				writeC("/app/direct-component", `
+resources:
+- ../base
+configMapGenerator:
+- name: my-configmap
+  behavior: merge
+  literals:	
+    - compValue=5
+    - testValue=2
+`),
+			},
+			runPath: "/app/direct-component",
+			expectedOutput: `
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: storefront
+spec:
+  replicas: 1
+---
+apiVersion: v1
+data:
+  compValue: "5"
+  otherValue: "10"
+  testValue: "2"
+kind: ConfigMap
+metadata:
+  annotations: {}
+  labels: {}
+  name: my-configmap-96dt22k28h
+`,
+		},
+		"missing-optional-component-api-version": {
+			input: []FileGen{writeTestBase, writeOverlayProd,
+				writeF("/app/comp/"+konfig.DefaultKustomizationFileName(), `
+kind: Component
+configMapGenerator:
+- name: my-configmap
+  behavior: merge
+  literals:	
+    - otherValue=9
+`),
+			},
+			runPath: "/app/prod",
+			expectedOutput: `
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: storefront
+spec:
+  replicas: 1
+---
+apiVersion: v1
+data:
+  otherValue: "9"
+  testValue: "1"
+kind: ConfigMap
+metadata:
+  annotations: {}
+  labels: {}
+  name: my-configmap-72cfg2mg5d
+---
+apiVersion: v1
+kind: Deployment
+metadata:
+  name: db
+spec:
+  type: Logical
+`,
+		},
+		// See how nameSuffix "-b" is also added to the resources included by "comp-a" because they are in the
+		// accumulator when "comp-b" is accumulated.  In practice we could use simple Kustomizations for this example.
+		"components-can-add-the-same-base-if-the-first-renames-resources": {
```

Comment:
this is an interesting but weird case.  I understand why it works this way, but found it a bit counter intuitive from a usage perspective.  consider adding a `TODO: figure out if this is the right behavior`